### PR TITLE
Findpeakbackground refactor 20796

### LINKED ISSUE #20796: Refactor FindPeakBackground
FindPeakBackground is heavily used by FitPeak, FindPeaks and future FitPeaks.  Calling it every time by creating the child algorithm by Algorithm factory is costly.  Thus FindPeakBackground shall be refactored such that a client method may use this functionality without creating a child algorithm.

### Expected behavior

A new class shall be created.  It will have the core functionality of FindPeakBackground.  And this class shall be used directly clients including FindPeakBackground.

### PR DESCRIPTION
**Pull request by wdzhou**
Refactor FindPeakbackground such that it can be used without using property manager.

**To test:**

1. Review the codes

2. Check whether all the tests are passed.

Fixes #20796 . 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/inc/MantidAlgorithms/FindPeakBackground.h [hunk 1]
```diff
@@ -65,6 +67,57 @@ class DLLExport FindPeakBackground : public API::Algorithm {
...
+  size_t m_iStartX; /// l0 in original code
+  size_t m_iStopX;  /// n in original code
+  /// background order: 0 for flat, 1 for linear, 2 for quadratic
+  size_t m_backgroundOrder;
+  /// constant sigma
+  double m_sigmaConstant;
+  /// output workspace (table of result)
+  API::ITableWorkspace_sptr m_outPeakTableWS;
+
+  size_t m_inputWSIndex;
```
Comment (quantumsteve) [inline Framework/Algorithms/inc/MantidAlgorithms/FindPeakBackground.h]: (minor)  This order seems odd to me. What would benefit from being in the first cache line?

**Review by quantumsteve — CHANGES_REQUESTED** [review]
Refactoring looks good to me and glad to see it doesn't change or break any tests.

I commented on changes that may improve this PR.

File: Framework/Algorithms/src/FindPeakBackground.cpp [hunk 2]
```diff
@@ -69,63 +69,87 @@ void FindPeakBackground::init() {
...
+  //  size_t n = sizey;
+  //  size_t l0 = 0;
+  //  if (m_vecFitWindows.size() > 1) {
+  //    Mantid::Algorithms::FindPeaks fp;
+  //    l0 = fp.getIndex(inpX, m_vecFitWindows[0]);
+  //    n = fp.getIndex(inpX, m_vecFitWindows[1]);
+  //    if (n < sizey)
+  //      n++;
+  //  }
+  //  // m_vecFitWindows won't be used again form this point till end.
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/Algorithms/src/FindPeakBackground.cpp]: will remove!

File: Framework/Algorithms/src/FindPeakBackground.cpp [hunk 3]
```diff
@@ -69,63 +69,87 @@ void FindPeakBackground::init() {
...
-  // Find background
+//----------------------------------------------------------------------------------------------
+/**
+ * @brief FindPeakBackground::addResultToOutput
+ * @param goodfit
+ * @param peakindexes
+ * @param bkgdparams
+ */
+void FindPeakBackground::addResultToOutput(
+    const int &goodfit, const std::vector<size_t> &peakindexes,
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/Algorithms/src/FindPeakBackground.cpp]: goodfit will be the returned value.

File: Framework/Algorithms/src/FindPeakBackground.cpp [hunk 4]
```diff
@@ -221,15 +247,24 @@ void FindPeakBackground::exec() {
...
     // Add a new row
-    API::TableRow t = m_outPeakTableWS->getRow(0);
-    t << static_cast<int>(inpwsindex) << static_cast<int>(min_peak)
-      << static_cast<int>(max_peak) << a0 << a1 << a2 << goodfit;
-  }
 
-  prog.report();
+    //    API::TableRow t = m_outPeakTableWS->getRow(0);
+    //    t << static_cast<int>(m_inputWSIndex) << static_cast<int>(min_peak)
+    //      << static_cast<int>(max_peak) << a0 << a1 << a2 << goodfit;
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/Algorithms/src/FindPeakBackground.cpp]: will remove

File: Framework/Algorithms/src/FindPeakBackground.cpp [hunk 5]
```diff
@@ -400,5 +435,145 @@ double FindPeakBackground::moment4(MantidVec &X, size_t n, double mean) {
...
+
+  setHistogram(inpWS->histogram(inpwsindex));
+
+  std::vector<double> fitwindow = getProperty("FitWindow");
+  setFitWindow(fitwindow);
+
+  // background
+  m_backgroundType = getPropertyValue("BackgroundType");
+  size_t bkgdorder = 0;
+  if (m_backgroundType.compare("Linear") == 0)
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/Algorithms/src/FindPeakBackground.cpp]: will apply

File: Framework/Algorithms/inc/MantidAlgorithms/FindPeakBackground.h [hunk 6]
```diff
@@ -65,6 +67,57 @@ class DLLExport FindPeakBackground : public API::Algorithm {
...
+  /// set fit window
+  void setFitWindow(const std::vector<double> &window);
+
+  /// add result to output workpsace
+  void addResultToOutput(const int &goodfit,
+                         const std::vector<size_t> &peakindexes,
+                         const std::vector<double> &bkgdparams);
+
+  /// histogram data to find peak background
+  boost::shared_ptr<const HistogramData::Histogram> m_histogram;
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/Algorithms/inc/MantidAlgorithms/FindPeakBackground.h]: HistogramData::Histogram& cannot be used directly as a class variable.

**Review by quantumsteve — APPROVED** [review]

**Pull request merged**